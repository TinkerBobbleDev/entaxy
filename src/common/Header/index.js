import React from 'react'
import AppBar from 'material-ui/AppBar'
import Toolbar from 'material-ui/Toolbar'
import Typography from 'material-ui/Typography'
import Button from 'material-ui/Button'
import { withStyles } from 'material-ui/styles'

const styles = {
  flex: {
    flex: 1,
  },
};

function Header(props) {
	const { classes } = props
	return (
		<AppBar position="static">
	    <Toolbar>
	      <Typography variant="title" color="inherit" className={classes.flex}>
	        Bean Counter
	      </Typography>
	      <Button color="inherit">Login</Button>
	    </Toolbar>
	  </AppBar>
  )
}

export default withStyles(styles)(Header)