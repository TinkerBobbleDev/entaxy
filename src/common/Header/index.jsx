import React from 'react'
import PropTypes from 'prop-types'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import { withStyles } from '@material-ui/core/styles'
import Logo from '../Logo/index'
import LoginButton from '../LoginButton'
import TopNav from '../TopNav'

const styles = {
  toolbar: {
    display: 'flex',
    'justify-content': 'space-between'
  }
}

const Header = ({ classes }) => (
  <AppBar position="static">
    <Toolbar className={classes.toolbar}>
      <Logo />
      <TopNav />
      <LoginButton />
    </Toolbar>
  </AppBar>
)

Header.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(Header)
