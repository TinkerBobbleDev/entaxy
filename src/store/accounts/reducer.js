/* eslint-disable no-case-declarations */
import types from './types'

export const initialState = {
  byId: {}, // {id1: {id: id1, name: 'account1', institution: 'TD'}, id2: {...}}
  byInstitution: {} // {TD: {accountIds: [id1, id2], balance: 100, apiKey: XYZ}, BMO: {...}}
}

export const groupByInstitution = ({ byId, byInstitution }) => {
  // Group sorted accounts by institution
  // results in something like:
  // {
  //   TD: {accountIds: [id1, id2], balance: 100},
  //   BMO: {accountIds: [id3], balance: 200}
  // }
  const groupedAccountIds = Object.values(byId)
    // Sort accounts by name
    .sort((a, b) => (a.name > b.name))
    // Group them by institution
    .reduce((result, account) => {
      const {
        id,
        institution,
        groupId,
        currentBalance
      } = account

      // Merge existing institution data
      const existingInstitutionData = byInstitution[institution] || { groups: {} }
      const currentInstitutionData = (result[institution] || { balance: 0, groups: {} })
      const newInstitutionData = { ...existingInstitutionData, ...currentInstitutionData }

      // Merge existing group data
      const existingGroupData = existingInstitutionData.groups[groupId] || {}
      const currentGroupData = currentInstitutionData.groups[groupId] || {
        id: groupId,
        balance: 0,
        accountIds: []
      }
      const newGroupData = { ...existingGroupData, ...currentGroupData }

      return {
        ...result,
        [institution]: {
          ...newInstitutionData,
          balance: currentInstitutionData.balance + currentBalance.localCurrency,
          groups: {
            ...newInstitutionData.groups,
            [groupId]: {
              type: 'default',
              ...newGroupData,
              balance: currentGroupData.balance + currentBalance.localCurrency,
              accountIds: [...newGroupData.accountIds, id]
            }
          }
        }
      }
    }, {})

  // But now the institutions are not sorted and we lost existing institution data
  return Object.keys(groupedAccountIds)
    // Sort institutions by name
    .sort((a, b) => (a > b))
    // Merge new sorted institutions with existing institution data
    .reduce((result, institution) => ({
      ...result,
      [institution]: {
        ...byInstitution[institution],
        ...groupedAccountIds[institution]
      }
    }), {})
}

let accounts
let institutionData
export default (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOAD_ACCOUNTS:
      return payload || initialState
    case types.CREATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.UPDATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.DELETE_ACCOUNT:
      const { [payload]: _, ...rest } = state.byId
      return {
        ...state,
        byId: rest
      }
    case types.GROUP_BY_INSTITUTION:
      return {
        ...state,
        byInstitution: groupByInstitution(state)
      }
    case types.CREATE_ACCOUNT_GROUP:
    case types.UPDATE_ACCOUNT_GROUP:
      if (payload === undefined) return state
      const { institution, accountGroup } = payload
      institutionData = state.byInstitution[institution] || { groups: {} }
      return {
        ...state,
        byInstitution: {
          ...state.byInstitution,
          [institution]: {
            ...institutionData,
            groups: {
              ...institutionData.groups,
              [accountGroup.id]: accountGroup
            }
          }
        }
      }
    default:
      return state
  }
}
