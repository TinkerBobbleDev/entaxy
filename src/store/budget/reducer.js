import types from './types'
import budgetCategories from '../../data/budgetCategories'

// https://projects.susielu.com
// http://repec.sowi.unibe.ch/stata/palettes/index.html
const defaultColours = [
  '#1f78b4', '#b2df8a', '#e31a1c',
  '#ff7f00', '#cab2d6', '#a6cee3',
  '#33a02c', '#6a3d9a', '#fb9a99',
  '#fdbf6f', '#ffff99', '#b15928'
]

export const initialState = (() => {
  const colours = {}
  let count = 0
  const categories = Object.keys(budgetCategories).sort().map(category => (
    {
      label: category,
      options: budgetCategories[category].map((subCategory) => {
        const colour = defaultColours[count % defaultColours.length]
        colours[subCategory] = colour
        count += 1
        return { label: subCategory, value: subCategory, colour }
      })
    }
  ))
  return {
    categories,
    colours,
    rules: {} // rules have the format => match: { category: 'cat 1', type: 'exact_match', count: 0 }
  }
})()

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_BUDGET:
      return action.payload || initialState
    case types.CREATE_EXACT_RULE:
      return {
        ...state,
        rules: {
          ...state.rules,
          [action.payload.match]: {
            category: action.payload.category,
            type: 'exact_match',
            count: 0
          }
        }
      }
    case types.DELETE_EXACT_RULE:
      // eslint-disable-next-line no-case-declarations
      const r = Object.keys(state.rules).reduce((result, match) => {
        if (match !== action.payload) {
          return {
            ...result,
            rules: { ...result.rules, [match]: state.rules[match] }
          }
        }
        return result
      }, { ...state, rules: {} })
      return r
    case types.COUNT_RULE_USAGE:
      // reset counters
      // eslint-disable-next-line no-case-declarations
      const newState = Object.keys(state.rules).reduce((result, match) => (
        {
          ...result,
          rules: {
            ...result.rules,
            [match]: { ...result.rules[match], count: 0 }
          }
        }
      ), state)

      // count the transactions
      for (const transaction of action.payload) {
        if (transaction.description in newState.rules) {
          newState.rules[transaction.description].count += 1
        }
      }
      // remove unused rules
      return Object.keys(newState.rules).reduce((result, match) => {
        if (newState.rules[match].count > 0) {
          return {
            ...result,
            rules: { ...result.rules, [match]: newState.rules[match] }
          }
        }
        return result
      }, { ...newState, rules: {} })
    default:
      return state
  }
}
