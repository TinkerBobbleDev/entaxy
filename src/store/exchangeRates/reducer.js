import { omit } from 'lodash'
import parse from 'date-fns/parse'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_EXCHANGE_RATES:
      return action.payload || initialState
    case types.UPDATE_EXCHANGE_RATES:
      // payload = {
      //   '2019-01-01': { GBP: 1.4, CAD: 1.0 },
      //   '2019-02-02': { GBP: 1.3}
      // }
      // eslint-disable-next-line no-case-declarations
      const newState = Object.keys(action.payload)
        .reduce((res1, date) => {
          return Object.keys(action.payload[date])
            .reduce((res2, currency) => ({
              ...res2,
              [currency]: {
                ...res2[currency],
                [parse(date).getTime()]: action.payload[date][currency]
              }
            }), res1)
        }, state)
      Object.keys(newState).forEach((currency) => {
        newState[currency].dates = Object.keys(newState[currency])
          .filter(c => c !== 'dates')
          .sort((a, b) => b - a)
      })
      return newState
    case types.DELETE_CURRENCIES:
      return omit(state, action.payload)
    default:
      return state
  }
}
