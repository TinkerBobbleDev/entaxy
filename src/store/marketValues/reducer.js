import _ from 'lodash'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_MARKET_VALUES:
      return action.payload || initialState
    case types.CREATE_MARKET_VALUE:
      return {
        ...state,
        [action.payload.ticker]: {
          marketValue: action.payload.marketValue,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.UPDATE_MARKET_VALUE:
      return {
        ...state,
        [action.payload.ticker]: {
          marketValue: action.payload.marketValue,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.DELETE_MARKET_VALUES:
      return _.omit(state, action.payload)
    default:
      return state
  }
}
