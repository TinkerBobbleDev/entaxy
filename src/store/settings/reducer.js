import _ from 'lodash'
import types from './types'

export const initialState = {
  portfolioFilters: {
    source: {},
    account: {}
  }
}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_SETTINGS:
      return action.payload || initialState
    case types.UPDATE_PORTFOLIO_FILTER_VALUE:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            [action.payload.option]: action.payload.value
          }
        }
      }
    case types.CREATE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            ...action.payload.options
          }
        }
      }
    case types.DELETE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: _.omit(state.portfolioFilters[action.payload.filterName], action.payload.options)
        }
      }
    default:
      return state
  }
}
