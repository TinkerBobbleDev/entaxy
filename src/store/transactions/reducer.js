import _ from 'lodash'
import types from './types'

export const initialState = []

export default (state = initialState, action) => {
  let index = null

  switch (action.type) {
    case types.LOAD_TRANSACTIONS:
      return action.payload || initialState
    case types.CREATE_TRANSACTION:
      return [...state, action.payload]
    case types.DELETE_TRANSACTION:
      index = _.findIndex(state, transaction => transaction.id === action.payload)
      return [...state.slice(0, index), ...state.slice(index + 1)]
    default:
      return state
  }
}
