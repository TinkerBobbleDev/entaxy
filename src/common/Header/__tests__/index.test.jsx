import React from 'react'
import { render, fireEvent, waitForElement } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { Provider } from 'react-redux'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import configureMockStore from 'redux-mock-store'
import Header from '..'
import { initialState as userInitialState } from '../../../store/user/reducer'

// https://github.com/mui-org/material-ui/issues/15726
global.document.createRange = () => ({
  setStart: () => {},
  setEnd: () => {},
  commonAncestorContainer: {
    nodeName: 'BODY',
    ownerDocument: document
  }
})

const renderContent = () => {
  const mockStore = configureMockStore()
  const store = mockStore({ user: userInitialState })
  const history = createMemoryHistory()

  const props = {
    match: { path: '/dashboard' }
  }
  return {
    ...render(
      <Provider store={store}>
        <Router history={history}>
          <Header {...props}>
            <div>content</div>
          </Header>
        </Router>
      </Provider>
    ),
    props,
    history
  }
}


describe('Header', () => {
  it('renders correctly', () => {
    const { getByText } = renderContent()

    expect(getByText('Dashboard')).toBeInTheDocument()
    expect(getByText('Budget')).toBeInTheDocument()
  })

  it('opens and closes menu', async () => {
    const { getByText, queryByText, history } = renderContent()
    const historyPushSpy = jest.spyOn(history, 'push')

    expect(queryByText('Money Flow')).not.toBeInTheDocument()
    expect(queryByText('History')).not.toBeInTheDocument()
    expect(queryByText('Categories')).not.toBeInTheDocument()
    // open the budget menu
    fireEvent.click(getByText('Budget'))
    waitForElement(() => getByText('Money Flow'))
    expect(getByText('History')).toBeInTheDocument()
    expect(getByText('Categories')).toBeInTheDocument()
    // close the menu
    fireEvent.click(getByText('Categories'))
    expect(historyPushSpy).toHaveBeenCalledWith('/budget-categories')
  })
})
