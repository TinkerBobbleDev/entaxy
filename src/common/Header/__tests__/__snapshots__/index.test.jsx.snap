// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<header
  className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-2 MuiAppBar-positionStatic-6 MuiAppBar-colorPrimary-8"
>
  <div
    className="MuiToolbar-root-37 MuiToolbar-gutters-38 Header-toolbar-1"
  >
    <div
      className="Logo-logo-39"
    >
      <img
        alt="Entaxy logo"
        className="Logo-image-40"
        src="logo.png"
      />
      <a
        className="MuiTypography-root-42 MuiTypography-title-48 MuiTypography-colorInherit-61 Logo-title-41"
        href="/"
        onClick={[Function]}
      >
        Entaxy
      </a>
    </div>
    <div />
    <button
      className="MuiButtonBase-root-91 MuiButton-root-67 MuiButton-text-69 MuiButton-flat-72 MuiButton-colorInherit-86"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-68"
      >
        Login
      </span>
      <span
        className="MuiTouchRipple-root-94"
      />
    </button>
  </div>
</header>
`;
